### ISSUE TITLE
Teach magit-git-command to complete arguments

### ISSUE THREAD

**Issue by tarsius** [magit/magit#2951]
Personally I almost never use `magit-git-command` and friends because they offer no completion, and instead I just drop to a terminal.  But this command is often recommended when someone asks how something can be done in Magit and it can't, as in "you can always use ! to enter arbitrary git commands".  Well yes, but this shouldn't be less nice in Magit than it is in a shell.  (Also we probably should provide the feature in question in some other way, but that's not the point here.)

The built-in `pcomplete` package provides something similar to "bash-completion", but like with that someone has to actually compile a "list" of possible completions.

The `pcmpl-git` package does that. It provides the function `pcomplete/git`: "Completion rules for the `git` command.".  It also contains code to automatically compile the lists from the asciidoc sources of the git man-pages.

The `git-command` package uses `pcmpl-git` to implement a command similar to `magit-git-command`, but it doesn't send the output to a Magit buffer.  Never-the-less we should also luck at this package, it might contain some insights.

**Comment by tarsius** [comment]
Also look at the discussion in #3004.

**Comment by tarsius** [comment]
> Teach magit-git-command to complete arguments

Actually, since there recently have been some attempts to improve `magit-git-command`, let me rephrase that: "do not provide magit-specific but inferior alternatives to `async-shell-command`.

For the time being this can rephrased as "do not try to incrementally improve `magit-git-command` et al., it is a lost cause.

The magit-specific variants have some features that the generic ones do not. Because of that we cannot just remove the former, some users will favor those features over those available in the generic variants. Also I don't think those features are useless, just significantly less important. They are:

1. Magit's status buffer is refreshed when the process returns.
2. Output goes into the "magit process" buffer.
3. Some variants save the user from having to type "git ".

If someone wants to tackle before I get around to it, then s/he has to produce something that is just as capable as `shell-command` and that preserves these three features.

---

Here are some other discussions about the subject. They do contain useful discussions. #3004, #3005,  #3007, #3041.

**Comment by npostavs** [comment]
So basically we just want to use a shell for `magit-git-command` as well?

**Comment by npostavs** [comment]
As in: https://github.com/magit/magit/compare/np/cmd-shell
(this drops `magit-git-global-arguments`, probably they should be kept somehow)

**Comment by npostavs** [comment]
Looks good to me. I'm a bit curious now as to what your original idea was though :)

**Comment by tarsius** [comment]
I am not sure anymore. I guess I was just confused by the scattered discussion. Anyway, I think this is a big improvement and just because things could be better still, we shouldn't delay this. I was a bit frustrated because the earlier proposals didn't simplify matters, instead they made things more complicated. But your approach gets the essential part right by throwing out the accidental complexity instead of piling on. I'll just leave it at that for the time being.
